crates/sui/src/client_commands.rs

```diff
@@ -840,6 +854,66 @@ impl SuiClientCommands {
                 };
                 SuiClientCommandResult::Addresses(output)
             }
+            SuiClientCommands::Balance {
+                address,
+                coin_type,
+                with_coins,
+            } => {
+                let address = get_identity_address(address, context)?;
+                let client = context.get_client().await?;
+                let mut objects: Vec<Coin> = Vec::new();
+                let mut cursor = None;
+
+                loop {
+                    let response = client
+                        .coin_read_api()
+                        .get_all_coins(address, cursor, None)
+                        .await?;
+                    objects.extend(response.data);
+
+                    if response.has_next_page {
+                        cursor = response.next_cursor;
+                    } else {
+                        break;
+                    }
+                }
+
+                let mut coins_by_type: BTreeMap<String, (Option<SuiCoinMetadata>, Vec<Coin>)> =
+                    BTreeMap::new();
+                for c in objects {
+                    match coins_by_type.entry(c.coin_type.clone()) {
+                        Entry::Occupied(mut e) => {
+                            e.get_mut().1.push(c);
+                        }
+                        Entry::Vacant(e) => {
+                            let coin_metadata = client
+                                .coin_read_api()
+                                .get_coin_metadata(c.coin_type.clone())
+                                .await
+                                .with_context(|| {
+                                    format!(
+                                        "Cannot fetch the coin metadata for coin {}",
+                                        c.coin_type
+                                    )
+                                })?;
+
+                            let _ = &e.insert((coin_metadata, vec![c]));
```

Comment:
Was there an issue with returning a mutable reference to the underlying vector? One of the benefits of the entry API is that you then don't have to repeat the post-initialisation logic (in this case, pushing `c`) in both branches.

Reply:
I was getting some errors from the borrow checker, so that was the way I found to make it work :D 
```
   Compiling sui v1.19.0 (/Users/src/sui/crates/sui)
error[E0596]: cannot borrow `e` as mutable, as it is not declared as mutable
   --> crates/sui/src/client_commands.rs:896:47
    |
896 |                         Entry::Occupied(e) => e.get_mut().1,
    |                                               ^ cannot borrow as mutable
    |
help: consider changing this to be mutable
    |
896 |                         Entry::Occupied(mut e) => e.get_mut().1,
    |                                         +++

error[E0507]: cannot move out of a mutable reference
   --> crates/sui/src/client_commands.rs:896:47
    |
896 |                         Entry::Occupied(e) => e.get_mut().1,
    |                                               ^^^^^^^^^^^^^ move occurs because value has type `Vec<sui_json_rpc_types::Coin>`, which does not implement the `Copy` trait
    |
help: consider borrowing here
    |
896 |                         Entry::Occupied(e) => &e.get_mut().1,
    |                                               +

error[E0507]: cannot move out of a mutable reference
   --> crates/sui/src/client_commands.rs:909:29
    |
909 | ...                   e.insert((coin_metadata, vec![])).1
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ move occurs because value has type `Vec<sui_json_rpc_types::Coin>`, which does not implement the `Copy` trait
    |
help: consider borrowing here
    |
909 |                             &e.insert((coin_metadata, vec![])).1
    |                             +

Some errors have detailed explanations: E0507, E0596.
For more information about an error, try `rustc --explain E0507`.
```

Reply:
This is as far as I could get.
```

error[E0597]: `e` does not live long enough
   --> crates/sui/src/client_commands.rs:896:56
    |
896 |                         Entry::Occupied(mut e) => &mut e.get_mut().1,
    |                                         -----          ^           - `e` dropped here while still borrowed
    |                                         |              |
    |                                         |              borrowed value does not live long enough
    |                                         binding `e` declared here

For more information about this error, try `rustc --explain E0597`.
warning: `sui` (lib) generated 1 warning
error: could not compile `sui` (lib) due to previous error; 1 warning emitted
```